File: archive/llm_written_v1/claude-sonnet-4.5/smith-waterman/java/test_smithwaterman.java
import java.security.MessageDigest;
import java.util.Arrays;
import java.nio.charset.StandardCharsets;

/**
 * Differential test harness for Smith-Waterman parallel implementation.
 * Tests correctness (vs sequential baseline), determinism, and performance.
 */
public class TestSmithWaterman {
    
    private static final int MATCH_SCORE = 3;
    private static final int MISMATCH_SCORE = -3;
    private static final int GAP_SCORE = -2;
    
    public static void main(String[] args) {
        boolean allPassed = true;
        
        System.out.println("=== Smith-Waterman Differential Testing ===\n");
        
        // Edge cases
        allPassed &= runTest("Edge: Empty query", "", "ACGT");
        allPassed &= runTest("Edge: Empty reference", "ACGT", "");
        allPassed &= runTest("Edge: Both empty", "", "");
        allPassed &= runTest("Edge: Single char match", "A", "A");
        allPassed &= runTest("Edge: Single char mismatch", "A", "T");
        
        // Small cases
        allPassed &= runTest("Small: Identical short", "ACGT", "ACGT");
        allPassed &= runTest("Small: Different short", "AAAA", "TTTT");
        allPassed &= runTest("Small: Partial match", "ACGTACGT", "CGTA");
        
        // Medium cases
        String medQuery = generateSequence(150, 12345);
        String medRef = generateSequence(200, 67890);
        allPassed &= runTest("Medium: Random 150x200", medQuery, medRef);
        
        String medQuery2 = "ACGT".repeat(50);
        String medRef2 = "TGCA".repeat(60);
        allPassed &= runTest("Medium: Repeated pattern 200x240", medQuery2, medRef2);
        
        // Large cases
        String largeQuery = generateSequence(500, 11111);
        String largeRef = generateSequence(600, 22222);
        allPassed &= runTest("Large: Random 500x600", largeQuery, largeRef);
        
        String largeQuery2 = generateSequence(800, 33333);
        String largeRef2 = generateSequence(700, 44444);
        allPassed &= runTest("Large: Random 800x700", largeQuery2, largeRef2);
        
        // Very large (triggers parallel path)
        String veryLargeQuery = generateSequence(1000, 55555);
        String veryLargeRef = generateSequence(1200, 66666);
        allPassed &= runTest("VeryLarge: Random 1000x1200", veryLargeQuery, veryLargeRef);
        
        // Determinism tests (run parallel twice)
        System.out.println("\n=== Determinism Tests ===\n");
        allPassed &= runDeterminismTest("Det: Medium 150x200", medQuery, medRef);
        allPassed &= runDeterminismTest("Det: Large 500x600", largeQuery, largeRef);
        allPassed &= runDeterminismTest("Det: VeryLarge 1000x1200", veryLargeQuery, veryLargeRef);
        
        // Performance comparison
        System.out.println("\n=== Performance Tests ===\n");
        runPerformanceTest("Perf: Large 500x600", largeQuery, largeRef);
        runPerformanceTest("Perf: Large 800x700", largeQuery2, largeRef2);
        runPerformanceTest("Perf: VeryLarge 1000x1200", veryLargeQuery, veryLargeRef);
        
        System.out.println("\n" + ("=".repeat(50)));
        if (allPassed) {
            System.out.println("✓ ALL TESTS PASSED");
            System.exit(0);
        } else {
            System.out.println("✗ SOME TESTS FAILED");
            System.exit(1);
        }
    }
    
    private static boolean runTest(String testName, String query, String reference) {
        System.out.println("Test: " + testName);
        System.out.println("  Query length: " + query.length() + ", Reference length: " + reference.length());
        
        try {
            SmithWatermanSequential seqSW = new SmithWatermanSequential(MATCH_SCORE, MISMATCH_SCORE, GAP_SCORE);
            SmithWaterman parSW = new SmithWaterman(MATCH_SCORE, MISMATCH_SCORE, GAP_SCORE);
            
            // Run sequential
            long startSeq = System.nanoTime();
            SmithWatermanSequential.AlignmentResult seqResult = seqSW.findAlignment(query, reference);
            long endSeq = System.nanoTime();
            
            // Run parallel
            long startPar = System.nanoTime();
            SmithWaterman.AlignmentResult parResult = parSW.findAlignment(query, reference);
            long endPar = System.nanoTime();
            
            // Compare results
            boolean scoreMatch = seqResult.score == parResult.score;
            boolean alignedAMatch = seqResult.alignedA.equals(parResult.alignedA);
            boolean alignedBMatch = seqResult.alignedB.equals(parResult.alignedB);
            boolean identityMatch = Math.abs(seqResult.identity - parResult.identity) < 0.0001;
            
            boolean passed = scoreMatch && alignedAMatch && alignedBMatch && identityMatch;
            
            double seqMs = (endSeq - startSeq) / 1_000_000.0;
            double parMs = (endPar - startPar) / 1_000_000.0;
            
            System.out.println("  Sequential: " + String.format("%.2f", seqMs) + " ms");
            System.out.println("  Parallel:   " + String.format("%.2f", parMs) + " ms");
            System.out.println("  Score match: " + scoreMatch + " (seq=" + seqResult.score + ", par=" + parResult.score + ")");
            System.out.println("  Alignment match: " + (alignedAMatch && alignedBMatch));
            System.out.println("  Result: " + (passed ? "✓ PASS" : "✗ FAIL"));
            System.out.println();
            
            return passed;
        } catch (Exception e) {
            System.out.println("  ✗ EXCEPTION: " + e.getMessage());
            e.printStackTrace();
            System.out.println();
            return false;
        }
    }
    
    private static boolean runDeterminismTest(String testName, String query, String reference) {
        System.out.println("Test: " + testName);
        System.out.println("  Query length: " + query.length() + ", Reference length: " + reference.length());
        
        try {
            SmithWaterman parSW = new SmithWaterman(MATCH_SCORE, MISMATCH_SCORE, GAP_SCORE);
            
            // Run parallel twice
            SmithWaterman.AlignmentResult result1 = parSW.findAlignment(query, reference);
            SmithWaterman.AlignmentResult result2 = parSW.findAlignment(query, reference);
            
            // Compute hashes
            String hash1 = computeHash(result1);
            String hash2 = computeHash(result2);
            
            boolean passed = hash1.equals(hash2);
            
            System.out.println("  Run 1 hash: " + hash1);
            System.out.println("  Run 2 hash: " + hash2);
            System.out.println("  Result: " + (passed ? "✓ DETERMINISTIC" : "✗ NON-DETERMINISTIC"));
            System.out.println();
            
            return passed;
        } catch (Exception e) {
            System.out.println("  ✗ EXCEPTION: " + e.getMessage());
            e.printStackTrace();
            System.out.println();
            return false;
        }
    }
    
    private static void runPerformanceTest(String testName, String query, String reference) {
        System.out.println("Test: " + testName);
        System.out.println("  Query length: " + query.length() + ", Reference length: " + reference.length());
        System.out.println("  Matrix size: " + (long)query.length() * reference.length());
        
        try {
            SmithWatermanSequential seqSW = new SmithWatermanSequential(MATCH_SCORE, MISMATCH_SCORE, GAP_SCORE);
            SmithWaterman parSW = new SmithWaterman(MATCH_SCORE, MISMATCH_SCORE, GAP_SCORE);
            
            // Warmup
            seqSW.findAlignment(query.substring(0, Math.min(50, query.length())), 
                               reference.substring(0, Math.min(50, reference.length())));
            parSW.findAlignment(query.substring(0, Math.min(50, query.length())), 
                               reference.substring(0, Math.min(50, reference.length())));
            
            // Sequential timing (3 runs)
            long seqTotal = 0;
            for (int i = 0; i < 3; i++) {
                long start = System.nanoTime();
                seqSW.findAlignment(query, reference);
                long end = System.nanoTime();
                seqTotal += (end - start);
            }
            double seqAvgMs = (seqTotal / 3.0) / 1_000_000.0;
            
            // Parallel timing (3 runs)
            long parTotal = 0;
            for (int i = 0; i < 3; i++) {
                long start = System.nanoTime();
                parSW.findAlignment(query, reference);
                long end = System.nanoTime();
                parTotal += (end - start);
            }
            double parAvgMs = (parTotal / 3.0) / 1_000_000.0;
            
            double speedup = seqAvgMs / parAvgMs;
            int cores = Runtime.getRuntime().availableProcessors();
            
            System.out.println("  Sequential avg: " + String.format("%.2f", seqAvgMs) + " ms");
            System.out.println("  Parallel avg:   " + String.format("%.2f", parAvgMs) + " ms");
            System.out.println("  Speedup:        " + String.format("%.2f", speedup) + "x");
            System.out.println("  Cores:          " + cores);
            System.out.println();
        } catch (Exception e) {
            System.out.println("  ✗ EXCEPTION: " + e.getMessage());
            e.printStackTrace();
            System.out.println();
        }
    }
    
    private static String generateSequence(int length, long seed) {
        char[] bases = {'A', 'C', 'G', 'T'};
        StringBuilder sb = new StringBuilder(length);
        long state = seed;
        for (int i = 0; i < length; i++) {
            // Simple LCG (linear congruential generator)
            state = (state * 1103515245L + 12345L) & 0x7fffffffL;
            sb.append(bases[(int)(state % 4)]);
        }
        return sb.toString();
    }
    
    private static String computeHash(SmithWaterman.AlignmentResult result) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            String data = result.alignedA + "|" + result.alignedB + "|" + 
                         result.score + "|" + result.identity;
            byte[] hash = md.digest(data.getBytes(StandardCharsets.UTF_8));
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }
            return hexString.toString().substring(0, 16);
        } catch (Exception e) {
            return "hash-error";
        }
    }
}

File: archive/llm_written_v1/claude-sonnet-4.5/smith-waterman/go/test_smith_waterman.go
package main

import (
	"crypto/sha256"
	"fmt"
	"os"
	"strings"
	"time"
)

func generateSequence(n int, pattern string) string {
	var sb strings.Builder
	for sb.Len() < n {
		sb.WriteString(pattern)
	}
	return sb.String()[:n]
}

func hashAlignment(alignedA, alignedB string, score int, identity float64) string {
	data := fmt.Sprintf("%s|%s|%d|%.6f", alignedA, alignedB, score, identity)
	hash := sha256.Sum256([]byte(data))
	return fmt.Sprintf("%x", hash)
}

type TestCase struct {
	name      string
	query     string
	reference string
}

func getTestCases() []TestCase {
	return []TestCase{
		{
			name:      "edge_empty",
			query:     "",
			reference: "",
		},
		{
			name:      "edge_one_empty",
			query:     "ACGT",
			reference: "",
		},
		{
			name:      "edge_single",
			query:     "A",
			reference: "A",
		},
		{
			name:      "small_exact_match",
			query:     "ACGTACGT",
			reference: "ACGTACGT",
		},
		{
			name:      "small_mismatch",
			query:     "ACGTACGT",
			reference: "TGCATGCA",
		},
		{
			name:      "small_partial",
			query:     "AGCTGACT",
			reference: "CGCTAGCT",
		},
		{
			name:      "medium_similar",
			query:     generateSequence(150, "ACGTACGT"),
			reference: generateSequence(150, "ACGTACGT"),
		},
		{
			name:      "medium_partial",
			query:     generateSequence(150, "AGCT"),
			reference: generateSequence(150, "CGTA"),
		},
		{
			name:      "large_similar",
			query:     generateSequence(500, "ACGTACGT"),
			reference: generateSequence(500, "ACGTACGT"),
		},
		{
			name:      "large_partial",
			query:     generateSequence(500, "AGCT"),
			reference: generateSequence(500, "CGTA"),
		},
	}
}

func main() {
	fmt.Println("=== Smith-Waterman Differential Test Suite ===\n")

	match := 2
	mismatch := -1
	gap := -1

	testCases := getTestCases()
	allPassed := true

	fmt.Println("Running correctness tests (sequential vs parallel)...")
	for _, tc := range testCases {
		// Sequential
		swSeq := NewSmithWatermanSeq(match, mismatch, gap)
		alignedA_seq, alignedB_seq, score_seq, identity_seq := swSeq.FindAlignment(tc.query, tc.reference)
		hashSeq := hashAlignment(alignedA_seq, alignedB_seq, score_seq, identity_seq)

		// Parallel
		swPar := NewSmithWaterman(match, mismatch, gap)
		alignedA_par, alignedB_par, score_par, identity_par := swPar.FindAlignment(tc.query, tc.reference)
		hashPar := hashAlignment(alignedA_par, alignedB_par, score_par, identity_par)

		passed := hashSeq == hashPar
		status := "✓ PASS"
		if !passed {
			status = "✗ FAIL"
			allPassed = false
		}

		fmt.Printf("  [%s] %s (q=%d, r=%d) - seq_hash=%s par_hash=%s\n",
			status, tc.name, len(tc.query), len(tc.reference), hashSeq[:8], hashPar[:8])

		if !passed {
			fmt.Printf("    SEQ: score=%d, identity=%.2f%%, alignA_len=%d, alignB_len=%d\n",
				score_seq, identity_seq, len(alignedA_seq), len(alignedB_seq))
			fmt.Printf("    PAR: score=%d, identity=%.2f%%, alignA_len=%d, alignB_len=%d\n",
				score_par, identity_par, len(alignedA_par), len(alignedB_par))
		}
	}

	fmt.Println("\nRunning determinism tests (parallel run twice)...")
	for _, tc := range testCases {
		// Skip tiny cases for determinism check
		if len(tc.query) < 10 {
			continue
		}

		swPar1 := NewSmithWaterman(match, mismatch, gap)
		alignedA1, alignedB1, score1, identity1 := swPar1.FindAlignment(tc.query, tc.reference)
		hash1 := hashAlignment(alignedA1, alignedB1, score1, identity1)

		swPar2 := NewSmithWaterman(match, mismatch, gap)
		alignedA2, alignedB2, score2, identity2 := swPar2.FindAlignment(tc.query, tc.reference)
		hash2 := hashAlignment(alignedA2, alignedB2, score2, identity2)

		passed := hash1 == hash2
		status := "✓ PASS"
		if !passed {
			status = "✗ FAIL"
			allPassed = false
		}

		fmt.Printf("  [%s] %s - run1=%s run2=%s\n", status, tc.name, hash1[:8], hash2[:8])
	}

	// Performance test on very large case
	fmt.Println("\nRunning performance test...")
	queryLarge := generateSequence(2000, "ACGTACGT")
	refLarge := generateSequence(2000, "ACGTACGT")

	swSeq := NewSmithWatermanSeq(match, mismatch, gap)
	startSeq := time.Now()
	_, _, scoreSeq, _ := swSeq.FindAlignment(queryLarge, refLarge)
	durationSeq := time.Since(startSeq).Seconds()

	swPar := NewSmithWaterman(match, mismatch, gap)
	startPar := time.Now()
	_, _, scorePar, _ := swPar.FindAlignment(queryLarge, refLarge)
	durationPar := time.Since(startPar).Seconds()

	speedup := durationSeq / durationPar
	fmt.Printf("  Large case (2000x2000): seq=%.4fs, par=%.4fs, speedup=%.2fx\n", durationSeq, durationPar, speedup)
	fmt.Printf("  Scores match: %v (seq=%d, par=%d)\n", scoreSeq == scorePar, scoreSeq, scorePar)

	// Write results to files
	summaryFile := "run_summary.txt"
	f, err := os.Create(summaryFile)
	if err != nil {
		fmt.Printf("Error creating summary file: %v\n", err)
		os.Exit(1)
	}
	defer f.Close()

	fmt.Fprintf(f, "Smith-Waterman Differential Test Results\n")
	fmt.Fprintf(f, "=========================================\n\n")
	fmt.Fprintf(f, "Correctness Tests:\n")
	for _, tc := range testCases {
		swSeq := NewSmithWatermanSeq(match, mismatch, gap)
		alignedA_seq, alignedB_seq, score_seq, identity_seq := swSeq.FindAlignment(tc.query, tc.reference)
		hashSeq := hashAlignment(alignedA_seq, alignedB_seq, score_seq, identity_seq)

		swPar := NewSmithWaterman(match, mismatch, gap)
		alignedA_par, alignedB_par, score_par, identity_par := swPar.FindAlignment(tc.query, tc.reference)
		hashPar := hashAlignment(alignedA_par, alignedB_par, score_par, identity_par)

		passed := hashSeq == hashPar
		status := "PASS"
		if !passed {
			status = "FAIL"
		}
		fmt.Fprintf(f, "  %s: %s (seq=%s, par=%s)\n", tc.name, status, hashSeq[:16], hashPar[:16])
	}

	fmt.Fprintf(f, "\nDeterminism Tests:\n")
	for _, tc := range testCases {
		if len(tc.query) < 10 {
			continue
		}

		swPar1 := NewSmithWaterman(match, mismatch, gap)
		alignedA1, alignedB1, score1, identity1 := swPar1.FindAlignment(tc.query, tc.reference)
		hash1 := hashAlignment(alignedA1, alignedB1, score1, identity1)

		swPar2 := NewSmithWaterman(match, mismatch, gap)
		alignedA2, alignedB2, score2, identity2 := swPar2.FindAlignment(tc.query, tc.reference)
		hash2 := hashAlignment(alignedA2, alignedB2, score2, identity2)

		passed := hash1 == hash2
		status := "PASS"
		if !passed {
			status = "FAIL"
		}
		fmt.Fprintf(f, "  %s: %s (run1=%s, run2=%s)\n", tc.name, status, hash1[:16], hash2[:16])
	}

	fmt.Fprintf(f, "\nPerformance Test:\n")
	fmt.Fprintf(f, "  Large case (2000x2000):\n")
	fmt.Fprintf(f, "    Sequential: %.4fs\n", durationSeq)
	fmt.Fprintf(f, "    Parallel:   %.4fs\n", durationPar)
	fmt.Fprintf(f, "    Speedup:    %.2fx\n", speedup)
	fmt.Fprintf(f, "    Scores match: %v (seq=%d, par=%d)\n", scoreSeq == scorePar, scoreSeq, scorePar)

	perfFile := "perf.txt"
	pf, err := os.Create(perfFile)
	if err != nil {
		fmt.Printf("Error creating perf file: %v\n", err)
		os.Exit(1)
	}
	defer pf.Close()

	fmt.Fprintf(pf, "Smith-Waterman Performance Results\n")
	fmt.Fprintf(pf, "==================================\n\n")
	fmt.Fprintf(pf, "Test Case: 2000x2000 sequences (N=4000000)\n")
	fmt.Fprintf(pf, "Sequential time: %.4fs\n", durationSeq)
	fmt.Fprintf(pf, "Parallel time:   %.4fs\n", durationPar)
	fmt.Fprintf(pf, "Speedup:         %.2fx\n", speedup)
	fmt.Fprintf(pf, "Scores match:    %v\n", scoreSeq == scorePar)

	fmt.Printf("\nResults written to %s and %s\n", summaryFile, perfFile)

	if !allPassed {
		fmt.Println("\n✗ Some tests FAILED")
		os.Exit(1)
	} else if speedup < 1.3 {
		fmt.Printf("\n✗ Performance gate not met: speedup %.2fx < 1.3x\n", speedup)
		os.Exit(1)
	} else {
		fmt.Println("\n✓ All tests PASSED")
		os.Exit(0)
	}
}
